src/shim.c

```diff
@@ -954,6 +966,54 @@ connect_to_proxy(struct cc_shim *shim)
 	return false;
 }
 
+inline void sleep_ms(int ms)
+{
+	struct timespec ts = { 0, ms * 1000000L };
+	nanosleep(&ts, NULL);
+}
+
+/*
+ * Try to re-establish tcp/unix connection with proxy in shim->timeout seconds.
+ *
+ * \param shim \ref cc_shim
+ *
+ * \return true on success, false on failure
+ */
+bool reconnect_to_proxy(struct cc_shim *shim)
+{
+	shim_warning("Reconnecting to cc-proxy (timeout %d s)\n",
+			shim->timeout);
+	int time = 0;
+	while (1) {
+		sleep_ms(RECONNECT_POLL_MS);
+		time += RECONNECT_POLL_MS;
+		if (time >= shim->timeout * 1000) {
+			shim_error("Failed to reconnect to cc-proxy "
+					"(timeout %d s)\n", shim->timeout);
+			return false;
+		}
+
+		close(shim->proxy_sock_fd);
+		if (! connect_to_proxy(shim)) {
+			continue;
+		}
+
+		shim_debug("Sending connect command\n");
+		if (! send_connect_command(shim)) {
```

Comment:
Since that's the second place in the code where we use both `connect_to_proxy()` with `send_connect_command()` right after (which is logical), I think we could factorize this through a new function.

Reply:
@dvoytik you have updated the PR, but I don't think you addressed this comment. Am I missing something ?

Reply:
@dvoytik ping ?